bliss/simulator/decoder.py

```diff
@@ -124,54 +161,55 @@ def render_images(self, tile_cat: TileCatalog, image_ids):
         psf_params = torch.stack(param_list, dim=0)
 
         # use the specified nmgy_to_nelec ratios indexed by image_id
-        nmgy_to_nelec_rats = [self.nmgy_to_nelec_dict[image_ids[b]] for b in range(batch_size)]
+        nmgy_to_nelec_rats = [self.physical_to_nelec_dict[image_ids[b]] for b in range(batch_size)]
 
         for b in range(batch_size):
             # Convert to electron counts
             tile_cat["star_fluxes"][b] *= nmgy_to_nelec_rats[b]
             tile_cat["galaxy_fluxes"][b] *= nmgy_to_nelec_rats[b]
 
         full_cat = tile_cat.to_full_params()
+
+        # generate random WCS shifts as manual image dithering via unaligning WCS
         wcs_batch = []
+        depth_shifts, wcs_for_depth = self.pixel_shifts(coadd_depth, 0)
 
         for b in range(batch_size):
             n_sources = int(full_cat.n_sources[b].item())
             psf = psfs[b]
-            shift, wcs = self.pixel_shift()
-            wcs_batch.append(wcs)
-            for band in range(self.n_bands):
-                gs_img = galsim.Image(array=images[b, band], scale=self.pixel_scale)
-                for s in range(n_sources):
-                    source_params = full_cat.one_source(b, s)
-                    source_type = source_params["source_type"].item()
-                    renderer = self.source_renderers[source_type]  # NOTE: SDSS-Specific!
-                    galsim_obj = renderer(psf, band, source_params)
-                    plocs0, plocs1 = source_params["plocs"]
-                    offset = np.array([plocs1 - (slen_w / 2), plocs0 - (slen_h / 2)])
-                    offset += shift[band]
-
-                    # essentially all the runtime of the simulator is incurred by this call
-                    # to drawImage
-                    galsim_obj.drawImage(
-                        offset=offset,
-                        method=self.psf_draw_method,
-                        add_to_image=True,
-                        image=gs_img,
+            band_shifts, band_wcs_list = self.pixel_shifts(self.n_bands, self.ref_band)
+            wcs_batch.append(band_wcs_list)
+            for d in range(coadd_depth):
+                for band in range(self.n_bands):
+                    band_img = galsim.Image(array=images[b, d, band], scale=self.pixel_scale)
+                    self.draw_sources_on_band_image(
+                        band_img,
+                        n_sources,
+                        full_cat,
+                        b,
+                        psf,
+                        band,
+                        image_dims=(slen_h, slen_w),
+                        band_shift=band_shifts[band],
+                        depth_shift=depth_shifts[d],
                     )
 
         # clamping here helps with an strange issue caused by galsim rendering
         images = torch.from_numpy(images).clamp(1e-8)
 
-        return images, psfs, psf_params, wcs_batch
+        return images, psfs, psf_params, wcs_batch, wcs_for_depth
```

Comment:
Would you add some more detailed documentation of the values returned by this function? People may not understand the distinction between `wcs_batch` and `wcs_for_depth`

Reply:
No longer needed with the new changes to unalignment